packages/base/src/commands.ts

```diff
@@ -51,6 +51,62 @@ function loadKeybindings(commands: CommandRegistry, keybindings: any[]) {
   });
 }
 
+function downloadFile(content: BlobPart, fileName: string, mimeType: string) {
+  const blob = new Blob([content], { type: mimeType });
+  const url = URL.createObjectURL(blob);
+  const downloadLink = document.createElement('a');
+  downloadLink.href = url;
+  downloadLink.download = fileName;
+  document.body.appendChild(downloadLink);
+  downloadLink.click();
+  document.body.removeChild(downloadLink);
+}
+
+function getSelectedLayer(
+  tracker: JupyterGISTracker
+): { model: IJupyterGISModel; selectedLayer: any; sources: any } | null {
+  const model = tracker.currentWidget?.model as IJupyterGISModel;
+  if (!model) {
+    return null;
+  }
+
+  const localState = model.sharedModel.awareness.getLocalState();
+  if (!localState || !localState['selected']?.value) {
+    return null;
+  }
+
+  const selectedLayerId = Object.keys(localState['selected'].value)[0];
+  const layers = model.sharedModel.layers ?? {};
+  const sources = model.sharedModel.sources ?? {};
+
+  const selectedLayer = layers[selectedLayerId];
+  if (!selectedLayer || !selectedLayer.parameters) {
+    return null;
+  }
+
+  return { model, selectedLayer, sources };
+}
+
+async function getGeoJSONString(
```

Comment:
Nitpicky, this could go in `tools.ts` and I'd rename it into

```suggestion
async function getGeoJSONDataFromLayerSource(
```